custom-contracts/nontransferable-zrc6/zrc6-nontransferable.scilla

```diff
@@ -0,0 +1,1051 @@
+  is_allowed = orb is_spender_or_operator is_token_owner;
+  match is_allowed with
+  | True =>
+  | False =>
+    error = NotAllowedToTransferError;
+    Throw error
+  end
+end
+
+procedure RequireIsAllowedToTransfer(token_owner: ByStr20, token_id: Uint256)
+
+  status <- is_transferable;
+  match status with
+  | False =>
+    (* Only contract owner can perform transfer  *)
+    RequireContractOwner
+  | True =>
+    (* Token owner, spender or operator can perform transfer *)
+    RequireAccessToTransfer token_owner token_id
+  end
+
+end
+
+procedure UpdateBalance(operation: Operation, address: ByStr20)
+  match operation with
+  | Add =>
+    maybe_count <- balances[address];
+    new_count = 
+      let cur_count = get_bal maybe_count in
+      (* if overflow occurs, it throws CALL_CONTRACT_FAILED *)
+      builtin add cur_count one;
+    balances[address] := new_count
+  | Sub =>
+    maybe_count <- balances[address];
+    new_count = 
+      let cur_count = get_bal maybe_count in
+      (* if underflow occurs, it throws CALL_CONTRACT_FAILED *)
+      builtin sub cur_count one;
+    balances[address] := new_count
+  end
+end
+
+(* @Requirements: *)
+(* - `to` must not be the zero address. Otherwise, it must throw `ZeroAddressDestinationError` *)
+(* - `to` must not be `_this_address`. Otherwise, it must throw `ThisAddressDestinationError` *)
+(* - `_sender` must be a minter. Otherwise, it must throw `NotMinterError` *)
+procedure MintToken(to: ByStr20)
+  RequireValidDestination to;
+
+  IsMinter _sender;
+
+  (* generate ID *)
+  current_token_id_count <- token_id_count;
+  new_token_id_count = builtin add current_token_id_count one;
+  token_id_count := new_token_id_count;
+  token_id = new_token_id_count;
+
+  (* mint a new token *)
+  token_owners[token_id] := to;
+
+  (* add one to the token owner balance *)
+  UpdateBalance add_operation to;
+  
+  (* add one to the total supply *)
+  current_supply <- total_supply;
+  new_supply = builtin add current_supply one;
+  total_supply := new_supply
+end
+
+procedure SetTokenURI(token_id: Uint256, token_uri: String)
+  is_empty_string = builtin eq token_uri empty_string;
+  match is_empty_string with 
+  | True => 
+    (* noop *)
+  | False =>
+    token_uris[token_id] := token_uri
+  end
+end
+
+procedure HandleMint(info: Pair ByStr20 String)
+  match info with
+  | Pair to token_uri =>
+    MintToken to;
+    token_id <- token_id_count;
+    SetTokenURI token_id token_uri
+  end
+end
+
+(* @Requirements: *)
+(* - `token_id` must exist. Otherwise, it must throw `TokenNotFoundError` *)
+(* - `_sender` must be a token owner or an operator. Otherwise, it must throw `NotOwnerOrOperatorError` *)
+procedure BurnToken(token_id: Uint256)
+  RequireExistingToken token_id;
+  maybe_token_owner <- token_owners[token_id];
+  match maybe_token_owner with
+  | None =>
+    (* Noop, never gets called *)
+  | Some token_owner =>
+    RequireOwnerOrOperator token_owner;
```

Comment:
@inamirandaz May need to restrict burn as well, good to add `RequireIsAllowedToTransfer` in BurnToken.

Reply:
@razill added, please check. Thanks